### PR TITLE
Box IPFS Upload

### PR DESCRIPTION
**Pull request by sparkplug0025**
Sorry for the huge PR ☺️

- Adds ipfs upload endpoint to the box
- Adds a configuration on eden webapp to use the box or not
- Adds flexible cid field parsing from the action data
- Adds a `common` package to share APIs and common functionalities between the box and the webapp

### CODE REVIEW

File: packages/box/src/handlers/ipfs-upload.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        );
+        if (parsedRequest.success !== true) {
+            throw new BadRequestError(parsedRequest.error.flatten());
+        }
+
+        const requestData: IpfsUploadRequest = parsedRequest.data;
+
+        if (!requestData.syncUpload) {
+            // We need to implement a background job mechanism to make sure the
+            // file can be upload asynchronously without blocking the client.
```
Comment (cool-ant) [inline packages/box/src/handlers/ipfs-upload.ts:42]: We should probably mark this obviously with a TODO so we can easily find this and be reminded to come back to it

File: packages/common/src/backend/error-handlers.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    if (error instanceof ServerError) {
+        return res
+            .status(error.status)
+            .json({ type: error.name, error: error.error });
+    } else {
+        return res
+            .status(500)
+            .json({ error: error.message || "Unknown error" });
+    }
+};
```
Comment (cool-ant) [inline packages/common/src/backend/error-handlers.ts:38]: So clean... very nicely done

File: packages/common/src/eos/ipfs-actions.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    try {
+        // Extract the CID supporting dot notation fields, eg: new_member_profile.img
+        const cidFields = validAction.cidField.split(".");
+        let cid = data;
+        while (cidFields.length) {
+            const cidField = cidFields.shift()!;
+            cid = cid[cidField];
+        }
+        return cid;
```
Comment (cool-ant) [inline packages/common/src/eos/ipfs-actions.ts:75]: I find this loop to be very hard to understand. Likely just that I don't know what the data looks like, but maybe you add a comment that explains or shows a tiny sample of data? I'm guessing cidFields is a dot-separated list of properties of an object, and this function is using cidFields to drill into the object to get the cid, which is nested? Some way of making that easier to get than re-reading this code would be great.

**Review by cool-ant — APPROVED** [review]
Just a couple minor comments. This all looks like super solid code. *Great* work.

File: packages/box/src/handlers/ipfs-upload.ts (hunk 1 above)
Comment (sparkplug0025) [reply packages/box/src/handlers/ipfs-upload.ts]: done! 👍

File: packages/common/src/eos/ipfs-actions.ts (hunk 3 above)
Comment (sparkplug0025) [reply packages/common/src/eos/ipfs-actions.ts:79]: that's exactly it. just added the comment! done! 👍

**Comment by brandonfancher** [comment]
@sparkplug0025 Progress. I've started reviewing this. Will finish tomorrow. I am noticing the build is failing (which you may already be aware of.) Sent you creds for a couple platforms for testing deployment of the docker containers too.

File: packages/box/src/handlers/ipfs-upload.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
+import { Request, Response } from "express";
+import multer from "multer";
```
Comment (brandonfancher) [inline packages/box/src/handlers/ipfs-upload.ts:2]: TIL of multer. Nice!

**Review by brandonfancher — APPROVED** [review]
This is awesome. Approved.

I do have a suggestion: The IPFS upload behavior and the box vs. infura options are not going to be immediately clear to other devs (internally _or_ those who might want to set up Eden at a different domain/elsewhere). I recommend adding:

1. to the webapp README a note about the two IPFS upload options and how to use them ((Infura + NextJS Serverless Pin + Pinata) vs. (Box + Pinata)) including env vars;
2. to the box README a note about any env vars that need to be set up.
3. to any package READMEs a note about how to whitelist contracts/actions.

Basically, just some light documentation.

But again, this is really nicely done. 👏 👏 👏

**Comment by sparkplug0025** [comment]
@brandonfancher great readme docs update suggestions! Working on it right now!
